[visionOS] Find overlay is missing behind find bar

pxlcoder opened the pull request:
#### 8b79cd47f0ad395d40922b70a9068c48a1c73c97
<pre>
[visionOS] Find overlay is missing behind find bar
<a href="https://bugs.webkit.org/show_bug.cgi?id=260815">https://bugs.webkit.org/show_bug.cgi?id=260815</a>
rdar://107376322

Reviewed by Wenson Hsieh.

On visionOS, the find bar appears in the scroll view&apos;s inset area. Currently,
WebKit does not display a find overlay in the inset area, as the overlay is
tied to the web content. Specifically, the find overlay is a `PageOverlay` in
the web process, since the overlay must appear behind the highlights (which are
drawn in Web process).

To fix, add another overlay layer to the scroll view, designed to cover the
inset area. The UI process-side overlay is inserted behind the content view, to
avoid a double overlay, while still covering the inset area. Additionally,
the overlay is sized to match the scroll view&apos;s bounds, and its position is
adjusted as the scroll view is scrolled, to avoid creating an unnecessarily
large layer. In order to synchronize the animation between the `PageOverlay`
originated layer, and the new UI process-originated layer, existing SPI is
leveraged, and the animation is fully driven from the UI process.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:]):
(-[WKWebView scrollViewDidScroll:]):
(-[WKWebView _frameOrBoundsMayHaveChanged]):
(-[WKWebView _animationForFindOverlay:]):
(-[WKWebView _updateFindOverlayPosition]):

Update the overlay whenever the scroll position or scroll view size changes.

(-[WKWebView _showFindOverlay]):

Some clients may attempt to show an overlay when one is already visible. Handle
that scenario by cancelling any existing animations.

(-[WKWebView _hideFindOverlay]):

Only remove layers after the animation is complete.

(-[WKWebView _didAddLayerForFindOverlay:]):
(-[WKWebView _didRemoveLayerForFindOverlay]):
(-[WKWebView _removeLayerForFindOverlay]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didEndTextSearchOperation): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView didBeginTextSearchOperation]):
(-[WKContentView didEndTextSearchOperation]):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::didEndTextSearchOperation): Deleted.
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didEndTextSearchOperation): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/267443@main">https://commits.webkit.org/267443@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/257bcdbe01d98c5ff192cfafc6231c000aca95fb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16430 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18206 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15408 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19982 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16894 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17773 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16626 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17053 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14207 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18975 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14299 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21687 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15285 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15056 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19366 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15646 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13287 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14849 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19218 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2049 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15471 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -191,8 +191,6 @@ void WebFoundTextRangeController::didBeginTextSearchOperation()
 
 void WebFoundTextRangeController::didEndTextSearchOperation()
```
Nit - can we delete this method altogether?

pxlcoder replied:
Yeah, removed this, as well as the plumbing around it. Thanks for the review!

pxlcoder committed:
[visionOS] Find overlay is missing behind find bar
https://bugs.webkit.org/show_bug.cgi?id=260815
rdar://107376322

Reviewed by Wenson Hsieh.

On visionOS, the find bar appears in the scroll view's inset area. Currently,
WebKit does not display a find overlay in the inset area, as the overlay is
tied to the web content. Specifically, the find overlay is a `PageOverlay` in
the web process, since the overlay must appear behind the highlights (which are
drawn in Web process).

To fix, add another overlay layer to the scroll view, designed to cover the
inset area. The UI process-side overlay is inserted behind the content view, to
avoid a double overlay, while still covering the inset area. Additionally,
the overlay is sized to match the scroll view's bounds, and its position is
adjusted as the scroll view is scrolled, to avoid creating an unnecessarily
large layer. In order to synchronize the animation between the `PageOverlay`
originated layer, and the new UI process-originated layer, existing SPI is
leveraged, and the animation is fully driven from the UI process.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:]):
(-[WKWebView scrollViewDidScroll:]):
(-[WKWebView _frameOrBoundsMayHaveChanged]):
(-[WKWebView _animationForFindOverlay:]):
(-[WKWebView _updateFindOverlayPosition]):

Update the overlay whenever the scroll position or scroll view size changes.

(-[WKWebView _showFindOverlay]):

Some clients may attempt to show an overlay when one is already visible. Handle
that scenario by cancelling any existing animations.

(-[WKWebView _hideFindOverlay]):

Only remove layers after the animation is complete.

(-[WKWebView _didAddLayerForFindOverlay:]):
(-[WKWebView _didRemoveLayerForFindOverlay]):
(-[WKWebView _removeLayerForFindOverlay]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didEndTextSearchOperation): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView didBeginTextSearchOperation]):
(-[WKContentView didEndTextSearchOperation]):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::didEndTextSearchOperation): Deleted.
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didEndTextSearchOperation): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/267443@main
